### PR TITLE
add FOV to camera model and slider popup

### PR DESCRIPTION
**Pull request by tlambert03**
# Description
There have been a couple requests to make camera perspective (aka "fov") a little more accessible/discoverable.
This PR adds a `fov` to the `viewer.camera` model, and a slider popup when you right click on the 2D/3D button (but currently only when you're in 3D mode ... not sure if that's good or confusion).

https://user-images.githubusercontent.com/1609449/116626524-5f1dd800-a919-11eb-85b1-62c6a83d25e1.mov

This isn't _that_ much more discoverable than the existing `Shift+RightClick+Drag` option (since it's unclear that you can right click it)...  but at least it prepares it in the model.

cc @sofroniewn 

## Type of change

- [ ] Bug-fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/docs/dev/guides/translations.html).

### CODE REVIEW

**Comment by alisterburt** [comment]
cool! cc from zulip, seeing some some problems with points points layers here (#2637)

**Review by jni — APPROVED** [review]

File: napari/_qt/widgets/qt_viewer_buttons.py
Comment (goanpeca) [inline napari/_qt/widgets/qt_viewer_buttons.py]: @tlambert03 how about addding this info to the tooltip?

>Right click to change the field of view of the camera when in 3D mode

Or similar

That would make it a little bit more discoverable?

File: napari/_qt/widgets/qt_viewer_buttons.py
Comment (goanpeca) [inline napari/_qt/widgets/qt_viewer_buttons.py]: Why not make this an instance method of QtLayerButtons @tlambert03 ?

**Review by goanpeca — COMMENTED** [review]
Thanks for working on this @tlambert03 

Have one suggestion and 1 question.

Thanks!

**Review by sofroniewn — COMMENTED** [review]
Exposing this functionality is great. I'd like to try out the PR tomorrow before merge just to see if I can find any weird corner cases not covered as somethings this stuff with vispy and changing cameras can be funny.

My only other question is if we have a better name than `fov`. I know that is what vispy uses, but find that very jargony, and I always think of something else when I hear it - to me the field of view has spatial units and is a diameter of what can be seen through a microscope (say the fov is 200um). Here the fov is in degrees and relates to an perspective projection.

I'm not quite sure what other name is better yet, but I'd like some discussion before merge. I'll try and think about it tomorrow too

**Comment by tlambert03** [comment]
> any weird corner cases

main corner case is that it totally fails with points, as @alisterburt pointed out.  (but the failure is no different than what would happen with shift Rmouse drag ... could set .fov back to zero if there are points?)

>  if we have a better name than fov. 

imaris just calls it perspective.  could do that?

**Comment by sofroniewn** [comment]
> main corner case is that it totally fails with points, as @alisterburt pointed out. (but the failure is no different than what would happen with shift Rmouse drag ... could set .fov back to zero if there are points?)

lol, ok maybe. I'll play around tomorrow and see what feels good.

> imaris just calls it perspective. could do that?

yes, i'd much prefer that!

**Comment by tlambert03** [comment]
did you want the model to also be `"perspective"`, or mostly just the slider in the gui?

**Comment by sofroniewn** [comment]
> did you want the model to also be "perspective", or mostly just the slider in the gui?

both please! I think `viewer.camera.perspective = 45` works, and consistency with gui and model is good

File: napari/_qt/widgets/qt_viewer_buttons.py
Comment (sofroniewn) [inline napari/_qt/widgets/qt_viewer_buttons.py]: These now need to be `perspective` too

File: napari/_qt/widgets/qt_viewer_buttons.py
Comment (sofroniewn) [reply napari/_qt/widgets/qt_viewer_buttons.py]: Thoughts @tlambert03? I don't have a preference.

Comment (tlambert03) [reply napari/_qt/widgets/qt_viewer_buttons.py]: moved

File: napari/_qt/widgets/qt_viewer_buttons.py
Comment (tlambert03) [reply napari/_qt/widgets/qt_viewer_buttons.py]: added

**Review by sofroniewn — APPROVED** [review]
Ok nice, i just played with this. It's looking good.

One thing right now though is that the `viewer.camera.perspective` can get out of sync if you change it using the shift right click drag method (which directly modifies the vispy camera fov). I think we should either disable that direct mouse function (and possibly reenable it via our own mouse functions) or see if we can connect a callback that allows us to stay in sync

**Comment by tlambert03** [comment]
> One thing right now though is that the `viewer.camera.perspective` can get out of sync if you change it using the shift right click drag method (which directly modifies the vispy camera fov).

thanks. fixed

**Review by goanpeca — APPROVED** [review]
🚀

**Comment by sofroniewn** [comment]
> > One thing right now though is that the viewer.camera.perspective can get out of sync if you change it using the shift right click drag method (which directly modifies the vispy camera fov).
>
> thanks. fixed

I can still get some funky things to happen, not quite sure why this occurred, but happened by trying to shift-drag the camera to change perspective in the cells example

```python-traceback
ERROR: Invoking <bound method Base3DRotationCamera.viewbox_mouse_event of <ArcballCamera at 0x7f96e0a7bd90>> for SceneMouseEvent
WARNING: Traceback (most recent call last):
  File "examples/add_image.py", line 12, in <module>
    napari.run()
  File "/Users/nsofroniew/GitHub/napari/napari/_qt/qt_event_loop.py", line 366, in run
    app.exec_()
  File "/Users/nsofroniew/GitHub/vispy/vispy/app/backends/_qt.py", line 526, in event
    out = super(QtBaseCanvasBackend, self).event(ev)
  File "/Users/nsofroniew/GitHub/vispy/vispy/app/backends/_qt.py", line 489, in mouseMoveEvent
    modifiers=self._modifiers(ev),
  File "/Users/nsofroniew/GitHub/vispy/vispy/app/base.py", line 213, in _vispy_mouse_move
    ev = self._vispy_canvas.events.mouse_move(**kwargs)
  File "/Users/nsofroniew/GitHub/vispy/vispy/util/event.py", line 454, in __call__
    self._invoke_callback(cb, event)
  File "/Users/nsofroniew/GitHub/vispy/vispy/util/event.py", line 470, in _invoke_callback
    cb(event)
  File "/Users/nsofroniew/GitHub/napari/napari/_vispy/vispy_canvas.py", line 76, in _process_mouse_event
    super()._process_mouse_event(event)
  File "/Users/nsofroniew/GitHub/vispy/vispy/scene/canvas.py", line 359, in _process_mouse_event
    getattr(picked.events, event.type)(scene_event)
  File "/Users/nsofroniew/GitHub/vispy/vispy/util/event.py", line 454, in __call__
    self._invoke_callback(cb, event)
  File "/Users/nsofroniew/GitHub/vispy/vispy/util/event.py", line 474, in _invoke_callback
    self, cb_event=(cb, event))
  << caught exception here: >>
  File "/Users/nsofroniew/GitHub/vispy/vispy/util/event.py", line 470, in _invoke_callback
    cb(event)
  File "/Users/nsofroniew/GitHub/vispy/vispy/scene/cameras/perspective.py", line 236, in viewbox_mouse_event
    self.scale_factor = self._event_value[0] * zoomy
TypeError: 'float' object is not subscriptable
```

Also if if try and make perspective really high with the drag then this happens

```python-traceback
WARNING: Traceback (most recent call last):
  File "examples/add_image.py", line 12, in <module>
    napari.run()
  File "/Users/nsofroniew/GitHub/napari/napari/_qt/qt_event_loop.py", line 366, in run
    app.exec_()
  File "/Users/nsofroniew/GitHub/vispy/vispy/app/backends/_qt.py", line 526, in event
    out = super(QtBaseCanvasBackend, self).event(ev)
  File "/Users/nsofroniew/GitHub/vispy/vispy/app/backends/_qt.py", line 489, in mouseMoveEvent
    modifiers=self._modifiers(ev),
  File "/Users/nsofroniew/GitHub/vispy/vispy/app/base.py", line 213, in _vispy_mouse_move
    ev = self._vispy_canvas.events.mouse_move(**kwargs)
  File "/Users/nsofroniew/GitHub/vispy/vispy/util/event.py", line 454, in __call__
    self._invoke_callback(cb, event)
  File "/Users/nsofroniew/GitHub/vispy/vispy/util/event.py", line 470, in _invoke_callback
    cb(event)
  File "/Users/nsofroniew/GitHub/napari/napari/_vispy/vispy_canvas.py", line 76, in _process_mouse_event
    super()._process_mouse_event(event)
  File "/Users/nsofroniew/GitHub/vispy/vispy/scene/canvas.py", line 359, in _process_mouse_event
    getattr(picked.events, event.type)(scene_event)
  File "/Users/nsofroniew/GitHub/vispy/vispy/util/event.py", line 454, in __call__
    self._invoke_callback(cb, event)
  File "/Users/nsofroniew/GitHub/vispy/vispy/util/event.py", line 474, in _invoke_callback
    self, cb_event=(cb, event))
  << caught exception here: >>
  File "/Users/nsofroniew/GitHub/vispy/vispy/util/event.py", line 470, in _invoke_callback
    cb(event)
  File "/Users/nsofroniew/GitHub/vispy/vispy/scene/cameras/perspective.py", line 264, in viewbox_mouse_event
    self.fov = min(180.0, max(0.0, fov))
  File "/Users/nsofroniew/GitHub/vispy/vispy/scene/cameras/base_camera.py", line 233, in fov
    raise ValueError("fov must be between 0 and 180.")
ValueError: fov must be between 0 and 180.
ERROR: Invoking <bound method Base3DRotationCamera.viewbox_mouse_event of <ArcballCamera at 0x7f96e0a7bd90>> for SceneMouseEvent
ERROR: Invoking <bound method Base3DRotationCamera.viewbox_mouse_event of <ArcballCamera at 0x7f96e0a7bd90>> repeat 2
```

We might want to put our own limits (possibly using pydantic on this (0 to 180)). Could come later though.

This stuff seems pretty edge case, but I wanted to flag in case others are seeing it too

**Comment by sofroniewn** [comment]
Just fixed a merge conflict, will merge once tests pass!

**Pull request merged**
